pkg/auth/auth_openshift.go

```diff
@@ -43,78 +45,70 @@ func validateAbsURL(value string) error {
 	return nil
 }
 
-func newOpenShiftAuth(ctx context.Context, c *openShiftConfig) (oauth2.Endpoint, *openShiftAuth, error) {
+func newOpenShiftAuth(ctx context.Context, k8sClient *http.Client, c *oidcConfig) (loginMethod, error) {
+	o := &openShiftAuth{
+		issuerURL:     c.issuerURL,
+		cookiePath:    c.cookiePath,
+		secureCookies: c.secureCookies,
+		k8sClient:     k8sClient,
+		getClient:     c.getClient,
+	}
+
+	// TODO: repeat the discovery several times as in the auth.go logic
+	var err error
+	o.oauthEndpointCache, err = NewAsyncCache[*oidcDiscovery](ctx, 5*time.Minute, o.getOIDCDiscoveryInternal)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct OAuth endpoint cache: %w", err)
+	}
+	o.oauthEndpointCache.Run(ctx)
+
+	return o, nil
+}
+
+func (o *openShiftAuth) getOIDCDiscovery() *oidcDiscovery {
+	return o.oauthEndpointCache.GetItem()
+}
+
+func (o *openShiftAuth) getOIDCDiscoveryInternal(ctx context.Context) (*oidcDiscovery, error) {
 	// Use metadata discovery to determine the OAuth2 token and authorization URL.
 	// https://access.redhat.com/documentation/en-us/openshift_container_platform/4.9/html/authentication_and_authorization/configuring-internal-oauth#oauth-server-metadata_configuring-internal-oauth
-	wellKnownURL := strings.TrimSuffix(c.issuerURL, "/") + "/.well-known/oauth-authorization-server"
+	wellKnownURL := strings.TrimSuffix(o.issuerURL, "/") + "/.well-known/oauth-authorization-server"
 
 	req, err := http.NewRequest(http.MethodGet, wellKnownURL, nil)
 	if err != nil {
-		return oauth2.Endpoint{}, nil, err
+		return nil, err
 	}
 
-	resp, err := c.k8sClient.Do(req.WithContext(ctx))
+	resp, err := o.k8sClient.Do(req.WithContext(ctx))
 	if err != nil {
-		return oauth2.Endpoint{}, nil, err
+		return nil, err
 	}
 	defer resp.Body.Close()
 
 	if resp.StatusCode/100 != 2 {
-		return oauth2.Endpoint{}, nil, fmt.Errorf("discovery through endpoint %s failed: %s",
+		return nil, fmt.Errorf("discovery through endpoint %s failed: %s",
 			wellKnownURL, resp.Status)
 	}
 
-	var metadata struct {
-		Issuer string `json:"issuer"`
-		Auth   string `json:"authorization_endpoint"`
-		Token  string `json:"token_endpoint"`
-	}
-
-	if err := json.NewDecoder(resp.Body).Decode(&metadata); err != nil {
-		return oauth2.Endpoint{}, nil, fmt.Errorf("discovery through endpoint %s failed to decode body: %v",
+	metadata := &oidcDiscovery{}
+	if err := json.NewDecoder(resp.Body).Decode(metadata); err != nil {
+		return nil, fmt.Errorf("discovery through endpoint %s failed to decode body: %v",
 			wellKnownURL, err)
 	}
 
-	if err := validateAbsURL(metadata.Issuer); err != nil {
-		return oauth2.Endpoint{}, nil, err
+	if metadata.Issuer != o.issuerURL {
+		return nil, fmt.Errorf("discovery provided unexpected issuer URL (%s)", metadata.Issuer)
```

Comment:
That's a good catch, I actually followed the spec and I did not realize our implementation was doing something slightly different. My bad completely :facepalm:

Reply:
The easiest solution is probably to just remove this check. I agree that this would be good to have, so we can open a bug or story to follow up with a more robust solution.

Reply:
I removed the check, my assumption was wrong as we're not actually doing a real OIDC discovery. I closed the related bug.